<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDA0321ENSkillsNetwork928-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="300" alt="Skills Network Logo"  />
    </a>
</p>


### Install the required libraries


In [2]:
!pip install requests
!pip install pillow

zsh:1: command not found: pip
zsh:1: command not found: pip


## Requests in Python


Requests is a Python Library that allows you to send <code>HTTP/1.1</code> requests easily. We can import the library as follows:


In [3]:
import requests

We will also use the following libraries:


In [4]:
import os 
from PIL import Image
from IPython.display import IFrame

You can make a <code>GET</code> request via the method <code>get</code> to [www.ibm.com](http://www.ibm.com/?utm_source=Exinfluencer&utm_content=000026UJ&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01&utm_medium=Exinfluencer&utm_term=10006555):


In [5]:

url='https://www.ibm.com/'
r=requests.get(url)


We have the response object <code>r</code>, this has information about the request, like the status of the request. You can view the status code using the attribute <code>status_code</code>.


In [6]:
r.status_code

200

You can view the request headers:


In [7]:
print(r.request.headers)

{'User-Agent': 'python-requests/2.32.3', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Cookie': '_abck=30ABA0D83C80CF1868B0C8F6A4150BAF~-1~YAAQl5UeuDVXHLuTAQAAmLXp8Q2BTmNfhlNpPXO45OdOriYKeI0sNRxVDOeHPtZRnLHamVCipsa/4124licdKWJ6/vUvDpoJzHmwBqbjUcwdCR5un2w6K54bHfrbyv1Q7K+Z8A1fEZ/lmrXQjnjCd0wVAUUTIpHWdMu7qjz05/lQ2/jQuiztZzCapaGMkpq81B+/5vFSKa1sQWdLcNYblB6tz4UHovyvwf3JDMjZ78fgcPIzQwLlaxP31JbxyDiwHA9XKwDF7pj9Pi5bctKEjbEAstdqaAQj7F55QlMftXkKZP3dzAd/JPUBzw4wD3+HM8xKuMG0qp8fdewUysbgezGH0nQAM8cCHhL+q2MpsWf2y+9qkc5pMRrxju+p9qSi9mId/IvEG8kJ0J+rWny7LagLXTtQVeU=~-1~-1~-1; bm_sz=0E0C2B658364B683D167F681342AC3F9~YAAQl5UeuDZXHLuTAQAAmLXp8Ro9uLvMGJkC48Tu5AqlnHT3pPqNyTb6pi7P9z+U02eEVgsKTXOKFBiYoGO1C6CYaPHyeWmf3E31YL8RlMUkHScajWu5PbYi/GkWXcFffijdHXRhqXoQo2kNsirb8Ils5DD1ElWwPvMI7P/6voYYbKiQhA0H+SzfbESr/RIXFfB9iYpt8+urV+motPpKAg2i+Jfu2WtGFXaJnCV7WEdKSxNm6I4fwsWnebK4TsZ1dqLRm7jS0iCfmpXIlU+me4yIyqsTg3jxYr95sKs66XXswDJDtxbqrFLfVWDnrTLgDTC48aIs1ScAwC6rLEfCCmRGiSBMLPL26U4Z~33

You can view the request body, in the following line, as there is nobody for a get request we get <code>None</code>:


In [8]:
print("request body:", r.request.body)

request body: None


You can view the <code>HTTP</code> response header using the attribute <code>headers</code>. This returns a python dictionary of <code>HTTP</code> response headers.


In [9]:
header=r.headers
print(r.headers)

{'Content-Security-Policy': 'upgrade-insecure-requests', 'x-frame-options': 'SAMEORIGIN', 'Last-Modified': 'Mon, 23 Dec 2024 04:06:58 GMT', 'ETag': 'W/"1ef18-629e820bf0c71-gzip"', 'Accept-Ranges': 'bytes', 'Content-Type': 'text/html;charset=utf-8', 'X-Content-Type-Options': 'nosniff', 'Cache-Control': 'max-age=900', 'Expires': 'Mon, 23 Dec 2024 05:22:25 GMT', 'X-Akamai-Transformed': '9 20198 0 pmb=mTOE,2', 'Content-Encoding': 'gzip', 'Date': 'Mon, 23 Dec 2024 05:07:25 GMT', 'Content-Length': '20411', 'Connection': 'keep-alive', 'Vary': 'Accept-Encoding', 'Strict-Transport-Security': 'max-age=31536000'}


You can obtain the date the request was sent using the key <code>Date</code>.


In [10]:
header['date']

'Mon, 23 Dec 2024 05:07:25 GMT'

<code>Content-Type</code> indicates the type of data:


In [11]:
header['Content-Type']

'text/html;charset=utf-8'

You can also check the <code>encoding</code>:


In [12]:
 r.encoding

'utf-8'

As the <code>Content-Type</code> is <code>text/html</code> you can use the attribute <code>text</code> to display the <code>HTML</code> in the body. You can review the first 100 characters:


In [13]:
r.text[0:100]

'\n<!DOCTYPE HTML>\n<html lang="en">\n<head>\r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n    \r\n      \r\n    \r\n  '

You can load other types of data for non-text requests, like images. Consider the URL of the following image:


In [14]:
# Use single quotation marks for defining string
url='https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/IDSNlogo.png'

You can make a get request:


In [15]:
r=requests.get(url)

You can look at the response header:


In [16]:
print(r.headers)

{'Date': 'Mon, 23 Dec 2024 05:07:43 GMT', 'X-Clv-Request-Id': '72a8304c-f2c1-4bfd-a6bb-327addc3b07b', 'Server': 'Cleversafe', 'X-Clv-S3-Version': '2.5', 'Accept-Ranges': 'bytes', 'x-amz-request-id': '72a8304c-f2c1-4bfd-a6bb-327addc3b07b', 'ETag': '"8bb44578fff8fdcc3d2972be9ece0164"', 'Content-Type': 'image/png', 'Last-Modified': 'Wed, 16 Nov 2022 03:32:41 GMT', 'Content-Length': '78776'}


You can see the <code>'Content-Type'</code>


In [17]:
r.headers['Content-Type']

'image/png'

An image is a response object that contains the image as a <a href="https://docs.python.org/3/glossary.html?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01#term-bytes-like-object">bytes-like object</a>. As a result, we must save it using a file object. First, you specify the <u>file path and
name</u>


In [18]:
path=os.path.join(os.getcwd(),'image.png')

You save the file, in order to access the body of the response we use the attribute <code>content</code> then save it using the <code>open</code> function and write <code>method</code>:


In [19]:
with open(path,'wb') as f:
    f.write(r.content)

You can view the image:


In [23]:
image.show()

<h3>Question: Download a file </h3>


Consider the following URL:


<code>URL = <https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt</code>


Write the commands to download the txt file in the given link.


In [24]:
## Write your code here
url = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0101EN-SkillsNetwork/labs/Module%205/data/Example1.txt'
path = os.path.join(os.getcwd(),'example1.txt')
r = requests.get(url)
with open(path,'wb') as f:
    f.write(r.content)

The Base URL is for <code>[http://httpbin.org/](http://httpbin.org/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01)</code>. It is a simple HTTP Request and Response service. The <code>URL</code> in Python is given by:


In [25]:
url_get='http://httpbin.org/get'

To create a Query string, add a dictionary. The keys are the parameter names and the values are the value of the Query string.


In [26]:
payload={"name":"Joseph","ID":"123"}

Then passing the dictionary <code>payload</code> to the <code>params</code> parameter of the <code> get()</code> function:


In [27]:
r=requests.get(url_get,params=payload)

You can print out the <code>URL</code> and see the name and values.


In [28]:
r.url

'http://httpbin.org/get?name=Joseph&ID=123'

There is no request body.


In [29]:
print("request body:", r.request.body)

request body: None


You can print out the status code.


In [30]:
print(r.status_code)

200


You can view the response as text:


In [31]:
print(r.text)

{
  "args": {
    "ID": "123", 
    "name": "Joseph"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Accept-Encoding": "gzip, deflate", 
    "Host": "httpbin.org", 
    "User-Agent": "python-requests/2.32.3", 
    "X-Amzn-Trace-Id": "Root=1-6768f128-10b5cf7c1057947733d310a1"
  }, 
  "origin": "104.28.123.19", 
  "url": "http://httpbin.org/get?name=Joseph&ID=123"
}



You can look at the <code>'Content-Type'</code>.


In [32]:
r.headers['Content-Type']

'application/json'

As the content <code>'Content-Type'</code> is in the <code>JSON</code> format, you can use the method <code>json()</code>, it returns a Python <code>dict</code>:


In [33]:
r.json()

{'args': {'ID': '123', 'name': 'Joseph'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.32.3',
  'X-Amzn-Trace-Id': 'Root=1-6768f128-10b5cf7c1057947733d310a1'},
 'origin': '104.28.123.19',
 'url': 'http://httpbin.org/get?name=Joseph&ID=123'}

The key <code>args</code> has the name and values:


In [34]:
r.json()['args']

{'ID': '123', 'name': 'Joseph'}

## Post Requests


Like a <code>GET</code> request, a <code>POST</code> is used to send data to a server, but the <code>POST</code> request sends the data in a request body. In order to send the Post Request in Python, in the <code>URL</code> you can change the route to <code>POST</code>:


In [35]:
url_post='http://httpbin.org/post'

This endpoint will expect data as a file or as a form. A form is convenient way to configure an HTTP request to send data to a server.


To make a <code>POST</code> request we use the <code>post()</code> function, the variable <code>payload</code> is passed to the parameter <code> data </code>:


In [36]:
r_post=requests.post(url_post,data=payload)

Comparing the URL from the response object of the <code>GET</code> and <code>POST</code> request you see the <code>POST</code> request has no name or value pairs.


In [37]:
print("POST request URL:",r_post.url )
print("GET request URL:",r.url)

POST request URL: http://httpbin.org/post
GET request URL: http://httpbin.org/get?name=Joseph&ID=123


You can compare the <code>POST</code> and <code>GET</code> request body, you see only the <code>POST</code> request has a body:


In [38]:
print("POST request body:",r_post.request.body)
print("GET request body:",r.request.body)

POST request body: name=Joseph&ID=123
GET request body: None


You can view the form as well:


In [39]:
r_post.json()['form']

{'ID': '123', 'name': 'Joseph'}

There is a lot more you can do. Check out <a href="https://requests.readthedocs.io/en/master/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Requests </a> for more.


## Authors

<p><a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01" target="_blank">Joseph Santarcangelo</a> 


### Other Contributors

<a href="https://www.linkedin.com/in/jiahui-mavis-zhou-a4537814a?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0101ENSkillsNetwork19487395-2021-01-01">Mavis Zhou</a>


<!--
## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description           |
| ----------------- | ------- | ---------- | ---------------------------- |
| 2024-10-04        | 2.5     | Madhusudan | ID reviewed                  |
| 2024-05-23        | 2.5     | Madhusudan | ID reviewed                  |
| 2023-11-02        | 2.4     | Abhishek Gagneja | Updated instructions   |
| 2023-06-07        | 2.3     |Akansha Yadav| Spell Check                 |
| 2021-12-20        | 2.1     | Malika     | Updated the links            |
| 2020-09-02        | 2.0     | Simran     | Template updates to the file |
|                   |         |            |                              |
|                   |         |            |                              |

## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>
--!>


Copyright © IBM Corporation. All rights reserved.
